exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginsHandlerService.java

```diff
@@ -0,0 +1,150 @@
+
+/**
+ * Drill plugins handler, which allows to update storage plugins configs from the
+ * {@link CommonConstants#STORAGE_PLUGINS_OVERRIDE_CONF} conf file
+ *
+ * TODO: DRILL-6564: It can be improved with configs versioning and service of creating
+ * {@link CommonConstants#STORAGE_PLUGINS_OVERRIDE_CONF}
+ */
+public class StoragePluginsHandlerService implements StoragePluginsHandler {
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(StoragePluginsHandlerService.class);
+
+  private final LogicalPlanPersistence lpPersistence;
+  private final DrillbitContext context;
+  private URL pluginsOverrideFileUrl;
+
+  public StoragePluginsHandlerService(DrillbitContext context) {
+    this.context = context;
+    this.lpPersistence = new LogicalPlanPersistence(context.getConfig(), context.getClasspathScan(),
+        new ObjectMapper(new HoconFactory()));
+  }
+
+  @Override
+  public void loadPlugins(@NotNull PersistentStore<StoragePluginConfig> persistentStore,
+                          @Nullable StoragePlugins bootstrapPlugins) {
+    // if bootstrapPlugins is not null -- fresh Drill set up
+    StoragePlugins pluginsForPersistentStore;
+
+    StoragePlugins newPlugins = getNewStoragePlugins();
+
+    if (newPlugins != null) {
+      pluginsForPersistentStore = new StoragePlugins(new HashMap<>());
+      Optional.ofNullable(bootstrapPlugins)
+          .ifPresent(pluginsForPersistentStore::putAll);
+
+      for (Map.Entry<String, StoragePluginConfig> newPlugin : newPlugins) {
+        String pluginName = newPlugin.getKey();
+        StoragePluginConfig oldPluginConfig = Optional.ofNullable(bootstrapPlugins)
+            .map(plugins -> plugins.getConfig(pluginName))
+            .orElse(persistentStore.get(pluginName));
+        StoragePluginConfig updatedStatusPluginConfig = updatePluginStatus(oldPluginConfig, newPlugin.getValue());
+        pluginsForPersistentStore.put(pluginName, updatedStatusPluginConfig);
+      }
+    } else {
+      pluginsForPersistentStore = bootstrapPlugins;
+    }
+
+    // load pluginsForPersistentStore to Persistent Store
+    Optional.ofNullable(pluginsForPersistentStore)
+        .ifPresent(plugins -> plugins.forEach(plugin -> persistentStore.put(plugin.getKey(), plugin.getValue())));
+
+    if (newPlugins != null) {
+      String fileAction = context.getConfig().getString(ACTION_ON_STORAGE_PLUGINS_OVERRIDE_FILE).toUpperCase();
+      Optional<ActionOnFile> actionOnFile = Arrays.stream(ActionOnFile.values())
+          .filter(action -> action.name().equals(fileAction))
+          .findFirst();
+      actionOnFile.ifPresent(action -> action.action(pluginsOverrideFileUrl));
+      // TODO: replace with ifPresentOrElse() once the project will be on Java9
+      if (!actionOnFile.isPresent()) {
+        logger.error("Unknown value {} for {} boot option", fileAction, ACTION_ON_STORAGE_PLUGINS_OVERRIDE_FILE);
```

Comment:
Also please indicate that nothing will be done with file.

Reply:
Done